plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryQueryPageSource.java

```diff
@@ -121,17 +124,27 @@ public BigQueryQueryPageSource(
         }, executor);
     }
 
-    private String buildSql(BigQueryTableHandle table, String projectId, List<BigQueryColumnHandle> columns, Optional<String> filter)
+    private String buildSql(BigQueryTableHandle table, String projectId, List<BigQueryColumnHandle> columns, Optional<String> filter, Optional<Long> limit)
     {
         if (isQueryFunction) {
             BigQueryQueryRelationHandle queryRelationHandle = (BigQueryQueryRelationHandle) table.relationHandle();
-            if (filter.isEmpty()) {
-                return queryRelationHandle.getQuery();
-            }
-            return "SELECT * FROM (" + queryRelationHandle.getQuery() + " ) WHERE " + filter.get();
+            return buildNativeQuery(queryRelationHandle.getQuery(), filter, limit);
         }
         TableId tableId = TableId.of(projectId, table.asPlainTable().getRemoteTableName().datasetName(), table.asPlainTable().getRemoteTableName().tableName());
-        return selectSql(tableId, ImmutableList.copyOf(columns), filter);
+        return selectSql(tableId, ImmutableList.copyOf(columns), filter, limit);
+    }
+
+    public static String buildNativeQuery(String nativeQuery, Optional<String> filter, Optional<Long> limit)
+    {
+        // projected column names can not be used for generating select sql because the query fails if it does not
+        // include a column name. eg: query => 'SELECT 1'
+        if (filter.isPresent()) {
+            String query = "SELECT * FROM (" + nativeQuery + ") WHERE " + filter.get();
+            return limit.map(value -> query + " LIMIT " + value)
+                    .orElse(query);
+        }
+        return limit.map(value -> "SELECT * FROM (" + nativeQuery + ") LIMIT " + value)
+                .orElse(nativeQuery);
```

Comment:
Limit application is same for both with and without filter - Can we unify them i.e `filter` can apply its cahcnages on top of the query and `LIMIT` can be applied on top of it - `SELECT * FROM (SELECT * FROM (...)) LIMIT 10` - is also a valid query right ?